### ISSUE TITLE
A bug when handling inheritance with `new` members

### ISSUE THREAD

**Issue by AlexeyRaga** [hedgehogqa/fsharp-hedgehog-experimental#71]
Hi, I have found one of the edge cases in which `GenX` fails to generate a correct instance, as well as the possible solution to the issue.

### Problem

Assuming that we have a hierarchy of two types:

```csharp
public abstract class A { public object Value { get; protected set; } }
public class B : A { public new string Value { get { return (string)base.Value; } set { base.Value = value; } } }
```

Notice that `B` has a property with the same name as `A`, but "hides" the base property with the `new` keyword.

Now, when we ask `GenX.auto<B>` then it generates an incorrect instance because `Value` property gets `obj` as a value and not `string`.

### Why it is happening

It is happening because this code:

```fsharp
| Shape.CliMutable (:? ShapeCliMutable<'a> as shape) ->
    let props = shape.Properties |> Seq.toList
```

gives us the list of _two_ properties in this order:

1. `Value` that is declared in `B`
2. `Value` that is declared in `A`

and values for these properties are generated and set _in this order_.
So first, `GenX` will generate a `correct` (by intention) value and will set it to `B.Value`, but _then_ it will generate an empty object and pass it into the "hidden" property, which overwrites the correct value with an incorrect one.

The list of properties comes to us from `TypeShape`, which just reflects all the instance properties in a given type:

```fsharp
and [<Sealed>] ShapeCliMutable<'Record> private (defaultCtor : ConstructorInfo) =
    let properties = lazy(
        typeof<'Record>.GetProperties(AllInstanceMembers)
        |> Seq.filter (fun p -> p.CanRead && p.CanWrite && p.GetIndexParameters().Length = 0)
        |> Seq.map (fun p -> mkWriteMemberUntyped<'Record> p.Name p [|p|])
        |> Seq.toArray)
```

`GetProperties` method explicitly does not specify the order in which properties are returned:

> The [GetProperties](https://docs.microsoft.com/en-us/dotnet/api/system.type.getproperties?view=net-6.0) method does not return properties in a particular order, such as alphabetical or declaration order. Your code must not depend on the order in which properties are returned, because that order varies.

so that we cannot rely on "this instance" property being passed last (and it does, indeed, not happen).

### Possible solution

Since we cannot rely on the order, one possible way of working around the issue would be to make sure that properties of "base instance" are set first, and properties of "this instance" are set last.

This would achieve that:

```fsharp
| Shape.CliMutable (:? ShapeCliMutable<'a> as shape) ->
    let selfProps, baseProps =
      shape.Properties
      |> Seq.toList
      |> List.partition (fun x -> x.MemberInfo.DeclaringType = typeof<'a>)

    (baseProps @ selfProps)  // <---- here is the change
    |> Seq.toList
    |> ListGen.traverse memberSetterGenerator
    |> Gen.map (fun fs -> fs |> List.fold (|>) (shape.CreateUninitialized ()))
```

and this way `GenX` will produce "correct by intention" value.

Of course, a "better" solution would be to detect `new` and to remove "hidden" members from the list, but:

1. I wasn't able to figure out how to do that reliably :)
2. Because types that we generated are blackboxes, it _may_ be logical to still initialise the base values with _something_, as we don't (and can't) know whether the child type getter/setter are accessing the base one or not.

So I think that the solution with setting "this instance" properties first should be good enough, and should definitely be better to what we currently have.

### Question

Will you accept a PR that fixes this issue in this way? ;)

**Comment by TysonMN** [comment]
Great work presenting a minimal example of this issue.  Yes, please do make a PR with your implementation.  I will also try to find a better fix.  In your PR, please add a failing test in one commit and put your fix in a second commit.

**Comment by AlexeyRaga** [comment]
Yes, will do.
I now think that probably a more "gold plated" solution from the same category would be to sort properties according to their hierarchy depth...  But if you find a better way then it'd be awesome.

**Comment by AlexeyRaga** [comment]
@TysonMN the PR is up.

**Comment by TysonMN** [comment]
Resolved by #72

**Comment by TysonMN** [comment]
First included in [version 0.7.0](https://www.nuget.org/packages/Hedgehog.Experimental/0.7.0)
